canopen/pdo.py

```diff
@@ -15,48 +12,97 @@
 
 logger = logging.getLogger(__name__)
 
+# TODO: Only send and receive messages when in state OPERATIONAL
+# TODO: Support the inhibit time constraint for transmission
+# TODO: Support more fine grained transmission types
+# TODO: Implement LocalPdoNode.export
+# TODO: We need to handle dummy mapping entries
+# TODO: Implement update_com_config
+# TODO: Implement update_map_config
+# TODO: Support more sophisticated trigger mechanisms
+# TODO: Support RTR trigger correctly
+# TODO: Invalid data changes shall be answered with an SDO abort
+
+
+def create_pdos(com_offset, map_offset, pdo_node, direction):
+    maps = {}
+    direction_lower = direction.lower()
+    if direction_lower in ["transmit", "tx"]:
+        PDO = TPDO
+    elif direction_lower in ["receive", "rx"]:
+        PDO = RPDO
+    else:
+        logger.critical("Direction %s is not supported" % direction)
+        return maps
+    for idx in range(0x200):
+        com_index = com_offset + idx
+        map_index = map_offset + idx
+        if com_index in pdo_node.node.object_dictionary:
+            com_entry = pdo_node.node.object_dictionary[com_index]
+            if com_entry.subindices[1].raw:
+                cob_id = com_entry.subindices[1].raw
+            else:
+                logger.warning("Don't know how to handle communication "
+                               "index 0x{:X}".format(com_index))
+                continue
+            maps[com_index] = PDO(cob_id, pdo_node, com_index, map_index)
+    return maps
 
-class PdoNode(collections.Mapping):
-    """Represents a slave unit."""
 
+class RemotePdoBase(collections.Mapping):
+    """Represents a slave unit."""
     def __init__(self, node):
         self.network = None
         self.node = node
-        self.rx = Maps(0x1400, 0x1600, self)
-        self.tx = Maps(0x1800, 0x1A00, self)
+        self.subscriptions = set()
+        self.rx = {}
+        self.tx = {}
 
     def __iter__(self):
-        for pdo_maps in (self.rx, self.tx):
-            for pdo_map in pdo_maps.values():
-                for var in pdo_map.map:
-                    yield var.name
+        return chain(iter(sorted(self.rx.keys())),
+                     iter(sorted(self.tx.keys())))
 
     def __getitem__(self, key):
-        for pdo_maps in (self.rx, self.tx):
-            for pdo_map in pdo_maps.values():
-                for var in pdo_map.map:
-                    if var.name == key:
-                        return var
+        if key in self.rx:
+            return self.rx[key]
+        if key in self.tx:
+            return self.tx[key]
         raise KeyError("%s was not found in any map" % key)
 
     def __len__(self):
-        count = 0
-        for pdo_maps in (self.rx, self.tx):
-            for pdo_map in pdo_maps.values():
-                count += len(pdo_map)
-        return count
-
-    def read(self):
```

Comment:
Keep `read` and `save` methods for backwards compatibility.

Reply:
Ok. I will fix this in a follow up commit